### ISSUE TITLE
Regression on pa11y 5.2.0 - URL fails to run

### ISSUE THREAD

**Issue by josebolos** [pa11y/pa11y#471]
The URL https://link.springer.com fails to run with pa11y 5.2.0:

```sh
npx pa11y@5.2.0 https://link.springer.com
Downloading Chromium r672088 - 108 Mb [====================] 100% 0.0s

Welcome to Pa11y

 > Running Pa11y on URL https://link.springer.com

Error: Evaluation failed: [object Object]
    at ExecutionContext._evaluateInternal (/Users/jose.bolos/.npm/_npx/49911/lib/node_modules/pa11y/node_modules/puppeteer/lib/ExecutionContext.js:122:13)
    at process._tickCallback (internal/process/next_tick.js:68:7)
  -- ASYNC --
    at ExecutionContext.<anonymous> (/Users/jose.bolos/.npm/_npx/49911/lib/node_modules/pa11y/node_modules/puppeteer/lib/helper.js:111:15)
    at DOMWorld.evaluate (/Users/jose.bolos/.npm/_npx/49911/lib/node_modules/pa11y/node_modules/puppeteer/lib/DOMWorld.js:112:20)
    at process._tickCallback (internal/process/next_tick.js:68:7)
  -- ASYNC --
    at Frame.<anonymous> (/Users/jose.bolos/.npm/_npx/49911/lib/node_modules/pa11y/node_modules/puppeteer/lib/helper.js:111:15)
    at Page.evaluate (/Users/jose.bolos/.npm/_npx/49911/lib/node_modules/pa11y/node_modules/puppeteer/lib/Page.js:782:43)
    at Page.<anonymous> (/Users/jose.bolos/.npm/_npx/49911/lib/node_modules/pa11y/node_modules/puppeteer/lib/helper.js:112:23)
    at runPa11yTest (/Users/jose.bolos/.npm/_npx/49911/lib/node_modules/pa11y/lib/pa11y.js:220:29)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

Pa11y 5.1.0 is able to run successfully on the same URL:

```sh
npx pa11y@5.1.0 https://link.springer.com
Downloading Chromium r672088 - 108 Mb [====================] 100% 0.0s

Welcome to Pa11y

 > Running Pa11y on URL https://link.springer.com

Results for URL: https://link.springer.com/

 • Error: This element has insufficient contrast at this conformance level. Expected a contrast ratio of at least 4.5:1, but text in this element has a contrast ratio of 4.32:1. Recommendation: change text colour to #3e7ab6.
   ├── WCAG2AA.Principle1.Guideline1_4.1_4_3.G18.Fail
   ├── #newest1 > noscript
   └── <noscript> <img src="https://st...</noscript>

[...]
```

Probable candidates for causing this are the updates to puppeteer and HTML_CodeSniffer.

**Comment by joeyciechanowicz** [comment]
This is most likely broken by the update to `HTML_CodeSniffer v2.3.0` which added:

> Support for alpha in colour contrast checks

**Comment by joeyciechanowicz** [comment]
Diggint into this it turns out this is caused by `HTML_CodeSniffer` using AMD if a `define` function is available.

The problem is that the build version provided via NPM incorrectly uglifies the module definition.
Theres a previous issue for this that was then closed: https://github.com/squizlabs/HTML_CodeSniffer/issues/192

We have a couple of options available for us.

### Temporarily disable AMD
We could remove AMD from the window, load the runner, then restore it.
```
let runnerBundle = `(function() {;
	const prevDefine = window.define;
	window.define = undefined;`;

	assertReporterCompatibility(runner, runnerModule.supports, pkg.version);

	for (const runnerScript of runnerModule.scripts) {
		runnerBundle += '\n\n';
		runnerBundle += await fs.readFile(runnerScript, 'utf-8');
	}

	runnerBundle += `;
	window.define = prevDefine;
	})();;`;
```

## Add support for AMD to runners
We could modify each runner to instead use `require` to load HTML_CS and execute it that way.
This would also require HTML_CS to be fixed so it provides a proper module.

**Comment by joeyciechanowicz** [comment]
Digging into this it turns out this is caused by `HTML_CodeSniffer` using AMD if a `define` function is available.

The problem is that the build version provided via NPM incorrectly uglifies the module definition.
Theres a previous issue for this that was then closed: https://github.com/squizlabs/HTML_CodeSniffer/issues/192

We have a couple of options available for us.

### Temporarily disable AMD
We could remove AMD from the window, load the runner, then restore it.
```
let runnerBundle = `(function() {;
	const prevDefine = window.define;
	window.define = undefined;`;

	assertReporterCompatibility(runner, runnerModule.supports, pkg.version);

	for (const runnerScript of runnerModule.scripts) {
		runnerBundle += '\n\n';
		runnerBundle += await fs.readFile(runnerScript, 'utf-8');
	}

	runnerBundle += `;
	window.define = prevDefine;
	})();;`;
```

## Add support for AMD to runners
We could modify each runner to instead use `require` to load HTML_CS and execute it that way.
This would also require HTML_CS to be fixed so it provides a proper module.

**Comment by imike57** [comment]
I tried to update html_codesniffer. But I still meet the same issue.
Something else to do ?

**Comment by josebolos** [comment]
I've reverted the change that was causing this behaviour and released it as [pa11y 5.2.1](https://github.com/pa11y/pa11y/releases/tag/5.2.1).

Please let us know if it doesn't fix your problem.
